### PR TITLE
pkgutil: add update all, check-mode, squashing and examples

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
Original PR #27866 from @scathatheworm

When working with Solaris pkgutil CSW packages, I came across this module being very basic in functionality, in particular, that I could not use it to update all CSW packages.

When going into details into the code I also found it did not incorporate a possibility of doing dry-run from the underlying utility, or supported to specify multiple packages for operations.

This module probably sees very little use, but it seemed like nice functionality to add and make it behave a little more like other package modules.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
pkgutil module

##### ANSIBLE VERSION

```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### ADDITIONAL INFORMATION
- Added ability to upgrade all packages:
```yaml
- pkgutil:
    name: '*'
    state: latest
```
- Added ability to modify state of a list of packages:
```yaml
- pkgutil:
    name:
    - CSWtop
    - CSWwget
    - CSWlsof
    state: present
```
- Added ability to have underlying tool perform a dry-run when using check mode, pkgutil -n

- Added ability to configure force option to force packages to state determined by repository (downgrade for example)
```yaml
- pkgutil:
    name: CSWtop
    state: latest
    force: yes
```
- Added more examples and documentation to show the new functionality

### CODE REVIEW

**Comment by dagwieers** [comment]
cc @albertomurillo @mavit @jtyr

**Comment by dagwieers** [comment]
I add integration tests, but they have not been tested (lacking any Solaris systems). Please test these changes by doing:
```bash
$ ansible-test integration --allow-unsupported --allow-destructive pkgutil -v
```
PS The integration tests are considered destructive because it may remove CWStop when not run successfully. So be aware.

File: test/sanity/validate-modules/ignore.txt [hunk 1]
```diff
@@ -761,6 +761,7 @@ lib/ansible/modules/packaging/os/pacman.py E326
 lib/ansible/modules/packaging/os/pkg5.py E326
 lib/ansible/modules/packaging/os/pkgin.py E322
 lib/ansible/modules/packaging/os/pkgng.py E322
+lib/ansible/modules/packaging/os/pkgutil.py E326
```
Comment (mattclay) [inline test/sanity/validate-modules/ignore.txt]: New ignores should not be added. The underlying issue should be corrected instead. If the issue is in existing module_utils or a docs fragment then the fix should be in a separate PR.

**Review by mattclay — DISMISSED** [review]

File: test/sanity/validate-modules/ignore.txt (hunk 1 above)
Comment (dagwieers) [reply test/sanity/validate-modules/ignore.txt]: The thing is that the documentation only offers **absent**, **latest** and **present** options for the state parameter, but (probably for backward compatibility) the argspec accepts also **installed** and **removed** options.

So we don't want to promote **installed* and **removed**, but I also don't want to break backward compatibility. Hence, ignoring the forced issue.

Comment (dagwieers) [reply test/sanity/validate-modules/ignore.txt]: BTW Beats me why the original module didn't have E326 though, it may be a bug in the testing?

Comment (mattclay) [reply test/sanity/validate-modules/ignore.txt]: @dagwieers The **installed** and **removed** state options are new in this PR, which is why the original module passes the tests.

Comment (dagwieers) [reply test/sanity/validate-modules/ignore.txt]: Oh, alright, they must have been added by the original author. I'll leave them out.
Sorry for the confusion.

**Comment by mavit** [comment]
> I added integration tests, but they have not been tested (lacking any Solaris systems). Please test these changes by doing:
> 
> ```shell
> $ ansible-test integration --allow-unsupported --allow-destructive pkgutil -v
> ```

I found that I also needed to pass `--no-temp-unicode` to get the tests to run on this Solaris 10 box.

File: lib/ansible/modules/packaging/os/pkgutil.py [hunk 2]
```diff
@@ -10,88 +10,140 @@
...
 
-def package_latest(module, name, site):
-    # Only supports one package
-    cmd = ['pkgutil', '-U', '--single', '-c']
+def packages_installed(module, names):
+    ''' Check if each package is installed and return list of the ones present '''
+    pkgs = []
+    for pkg in names:
+        if pkg.startswith('CSW'):
+            continue
```
Comment (mavit) [inline lib/ansible/modules/packaging/os/pkgutil.py:120]: This looks like debugging code that should have been removed?  It causes the module to fail to detect installed OpenCSW packages.

All tests succeed iff these two lines are removed.

Comment (dagwieers) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: @mavit I assumed because there is a better tool to manage OpenCSW packages ?

Comment (mavit) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: No, `pkgutil` is the OpenCSW package manager.  These are typically the only packages that it manages.

Comment (dagwieers) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: Ok, I see what the problem is, it should only consider CSW packages.
I'll fix.

Comment (dagwieers) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: Fixed.

**Comment by dagwieers** [comment]
@mavit Can you test again ? Does the integration test work out-of-the-box ? I wasn't expecting it to be perfect, but I cannot test myself.

**Comment by mattclay** [comment]
Any tests that require `--no-temp-unicode` to pass indicates there is a bug in the test and/or in Ansible itself that needs to be fixed. If you need help figuring out what the issue is then include details of the failure on this PR and I should be able to help.

File: lib/ansible/modules/packaging/os/pkgutil.py (hunk 2 above)
Comment (mavit) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: This leads to another issue.  The `pkgutil` command will quietly prefix package names with `CSW` if absent, and users might reasonably expect this Ansible module to behave in the same way.

Consider the following play, which will always return `changed`, `ok`, successfully installing the package but then neglecting to remove it.
```yaml
- pkgutil:
    name: top
    state: present

- pkgutil:
    name: top
    state: absent
```

It would be better to throw an error in this case rather than silently considering such packages never to be installed.

Comment (dagwieers) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: Look, I can't fix it because I cannot test this code.

Comment (dagwieers) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: (And mind you, I didn't wrote it either ;-))

Comment (mavit) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: Well, I suppose you *could* install Solaris in a VM, but I do see why you’d consider than the problem of someone who cares enough about Solaris to already have some Solaris.  That’s cool.  Frankly I’m impressed you tried to take this on in the first place.

So, then, let’s merge what we have.  The tests all pass, now, and it’s a step in the right direction.

Comment (dagwieers) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: Well, it usually starts small. And before you know it, it's refactored ;-)
If you are dedicated to test this module and improve the integration test, I prefer to merge and let you finish the job. Deal ?

Comment (mavit) [reply lib/ansible/modules/packaging/os/pkgutil.py:120]: Deal.

**Comment by mkrizek** [comment]
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general.

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project
